libphobos/libdruntime/object.d

```diff
@@ -1225,10 +1213,21 @@ class TypeInfo_Struct : TypeInfo
 
   @safe pure nothrow
   {
-    size_t   function(in void*)           xtoHash;
-    bool     function(in void*, in void*) xopEquals;
-    int      function(in void*, in void*) xopCmp;
-    string   function(in void*)           xtoString;
+    size_t          function(in void*)           xtoHash;
+    /* The xopEquals and xopCmp function pointers usually point to the struct's
+     * opEquals and opCmp methods. If the method doesn't take its single
+     * argument by reference, the front-end injects a static __xopEquals/
+     * __xopCmp function (taking 2 arguments, lhs `p` and rhs `q`).
+     *
+     * In the method case, lhs `p` is the `this` argument and must be passed
+     * as first argument before rhs `q`.
+     * Enforce this arguments order by marking the pointed-to functions as
+     * using the C calling convention, for which the arguments are never
+     * reversed (contrary to `extern (D)`).
+     */
+    extern (C) bool function(in void*, in void*) xopEquals;
```

Comment:
`xopEquals` is now in the global namespace. I wonder whether we should prefix all (new) `extern(C)` druntime functions with something?

Reply:
I don't think the frontend cares about that. As it makes the call through a function pointer, so this only changes ABI of how parameters are expected to be passed.

I didn't check the code in clone.d, but when I spoke to Daniel about it, I recall him saying that it's just an abi hack. The compiler sees xopCmp as a C function internally, but reverses the arguments as its really calling a D method.

This reversing of arguments never worked for gdc.

...

However I should check something about this, as it appears that the test for xopCmp parameter order was removed or no longer uses the generated xopCmp call.

Reply:
Oh, you're right, I totally forgot that this is just a pointer to a generated function (which then obviously has a unique name) 👍 
Does the frontend emit these `xopCmp` functions as `extern(C)` or `extern(D)`?

Reply:
That is just what I've pushed a test for.  The `__xopCmp` function is marked as LINKd at least.

Reply:
Although, now I think about it, it doesn't really matter for us as LINKd and LINKc are the same...